File: UsingTypeclasses/src/main/kotlin/typeclasses/JsonProgram.kt
package typeclasses

import typeclasses.marshalling.JsonMarshalling

fun main() {
    val person1 = Person("Jane")
    val order1 = Order(123.45)

    JsonMarshalling.forPerson().run {
        val text1 = person1.marshal()
        println(text1)

        val person2 = person1.unmarshal("{ \"person\": \"Dave\" }")
        println(person2)
    }

    JsonMarshalling.forOrder().run {
        val text2 = order1.marshal()
        println(text2)

        val order2 = order1.unmarshal("{ \"order\": 456.78 }")
        println(order2)
    }
}

File: UsingTypeclassesScala/src/main/scala/typeclasses/JsonProgram.scala
package typeclasses

import typeclasses.marshalling.Marshalling._
import typeclasses.marshalling.JsonMarshallingInstances._

object JsonProgram {
  def main(args: Array[String]): Unit = {
    val person1 = new Person("Jane")
    val order1 = new Order(123.45)


    val text1 = person1.marshal
    println(text1)

    val person2 = person1.unmarshal("{ \"person\": \"Dave\" }")
    println(person2)


    val text2 = order1.marshal
    println(text2)

    val order2 = order1.unmarshal("{ \"order\": 456.78 }")
    println(order2)
  }
}
